chore: Add unit tests for `pkg/apply`

akagami-harsh opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, check our contributor guidelines: https://www.kubeflow.org/docs/about/contributing
2. To know more about Kubeflow Trainer, check the developer guide:
    https://github.com/kubeflow/trainer/blob/master/CONTRIBUTING.md
3. If you want *faster* PR reviews, check how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** _(optional, in `Fixes #<issue number>, #<issue number>, ...` format, will close the issue(s) when PR gets merged)_:
- Fixes #2452


**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/trainer/) included if any changes are user facing

tenzen-y reviewed (commented):
Thanks, this is very initial comments.
Could you re-implement based on coding guide?
http://github.com/kubeflow/katib/blob/master/CONTRIBUTING.md#go-development

arnabnandikgp commented:
hey @akagami-harsh do you have the bandwidth to complete the work on this, or perhaps I can take this up in a new PR. Please let me know :)

andreyvelich commented:
Hi @akagami-harsh, are you still working on this or we can assign it to new contributor ?

akagami-harsh commented:
I wasn't able to work on it recently, but I'm still interested and will try complete it within this week.

astefanutti commented on pkg/apply/apply_test.go:
```diff
@@ -0,0 +1,337 @@
+	}
+}
+
+func TestEnvVar(t *testing.T) {
```
nit: would be nice to move it closer to `TestEnvVars`.

astefanutti commented on pkg/apply/apply_test.go:
```diff
@@ -0,0 +1,337 @@
+				*corev1ac.Volume().WithName("data"),
+			},
+			toUpsert: []corev1ac.VolumeApplyConfiguration{
+				*corev1ac.Volume().WithName("data").WithEmptyDir(corev1ac.EmptyDirVolumeSource()),
```
nit: break lines to keep the line short.

astefanutti reviewed (commented):
I've left a couple of nits, otherwise it looks good to me.

akagami-harsh commented:
Thanks for the review, @astefanutti! I've addressed your comments, refactored the tests, added additional test cases, and included tests for the `FromTypedObjWithFields` function as well

astefanutti commented:
/ok-to-test

akagami-harsh commented:
> @akagami-harsh Please can you rebase your PR, so we can review it ?

@andreyvelich done!

astefanutti commented:
/lgtm

Thanks!

andreyvelich reviewed (commented):
Thank you @akagami-harsh!
/cc @tenzen-y @Electronic-Waste

akagami-harsh commented on pkg/apply/apply_test.go:
```diff
@@ -0,0 +1,662 @@
+				*corev1ac.ContainerPort().WithContainerPort(9090),
+			},
+			want: []corev1ac.ContainerPortApplyConfiguration{
+				*corev1ac.ContainerPort().WithContainerPort(9090),
```
Because the upsert logic treats unnamed ports as the same “slot.” The match predicate considers ports equal if either the number matches or the name matches; with both names nil, name comparison is true and the new port replaces the old one.
```go
func byContainerPortOrName(a, b corev1ac.ContainerPortApplyConfiguration) bool {
	return ptr.Equal(a.ContainerPort, b.ContainerPort) || ptr.Equal(a.Name, b.Name)
}
```

andreyvelich replied:
Make sense!

akagami-harsh commented on pkg/apply/apply_test.go:
```diff
@@ -0,0 +1,662 @@
+				result, err := FromTypedObjWithFields[[]interface{}](tc.input, tc.fields...)
+				if err != nil {
+					t.Errorf("Unexpected error: %v", err)
+					return
```
Yes, one of them is necessary: the return at the end of the error-expected branch prevents falling through into the success-path assertions below. removed unnecessary ones

akagami-harsh commented on pkg/apply/apply_test.go:
```diff
@@ -0,0 +1,662 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestUpsertEnvVar(t *testing.T) {
```
Done!, consolidated it and simplified the duplicate test by collapsing `TestUpsertEnvVars` into a minimal smoke test

andreyvelich commented on pkg/apply/apply.go:
```diff
@@ -41,9 +41,7 @@ func UpsertEnvVar(envVars *[]corev1ac.EnvVarApplyConfiguration, envVar ...corev1
-	for _, e := range upEnvVars {
-		upsert(envVars, e, byEnvVarName)
-	}
+	UpsertEnvVar(envVars, upEnvVars...)
```
@akagami-harsh Can we just remove the `UpsertEnvVar()` function and use `UpsertEnvVars()` everywhere ?
Including this part: https://github.com/kubeflow/trainer/blob/9dd455e82228a8f40d6364cfcafc8ddd17f91e77/pkg/runtime/core/trainingruntime.go#L225-L227

kramaranya commented:
/ok-to-test

kramaranya commented:
Thank you for this, @akagami-harsh!! This looks great!
/lgtm
